scalding-core/src/main/scala/com/twitter/scalding/typed/CascadingBackend.scala

```diff
@@ -0,0 +1,268 @@
+package com.twitter.scalding.typed
+
+import cascading.flow.FlowDef
+import cascading.pipe.{ Each, Pipe }
+import cascading.tuple.{ Fields, TupleEntry }
+import com.twitter.scalding.TupleConverter.{ singleConverter, tuple2Converter }
+import com.twitter.scalding.TupleSetter.{ singleSetter, tup2Setter }
+import com.twitter.scalding._
+import com.twitter.scalding.serialization.{ CascadingBinaryComparator, OrderedSerialization, Boxed }
+import scala.collection.mutable.{ Map => MMap }
+import java.util.WeakHashMap
+
+object CascadingBackend {
+  import TypedPipe._
+
+  /**
+   * we want to cache renderings of some TypedPipe to Pipe so cascading
+   * will see them as the same. Without this, it is very easy to have
+   * a lot of recomputation. Ideally we would plan an entire graph
+   * at once, and not need a static cache here, but currently we still
+   * plan one TypedPipe at a time.
+   */
+  private[this] val pipeCache = new WeakHashMap[TypedPipe[Any], Map[Mode, CascadingPipe[Any]]]()
+
+  private def cacheGet[T](t: TypedPipe[T], m: Mode)(p: FlowDef => CascadingPipe[T]): CascadingPipe[T] = {
+    def add(mmc: Map[Mode, CascadingPipe[Any]]): CascadingPipe[T] = {
+      val emptyFD = new FlowDef
+      val res = p(emptyFD)
+      pipeCache.put(t, mmc + (m -> res.asInstanceOf[CascadingPipe[Any]]))
+      res
+    }
+
+    pipeCache.synchronized {
+      pipeCache.get(t) match {
+        case null => add(Map.empty)
+        case somemap if somemap.contains(m) => somemap(m).asInstanceOf[CascadingPipe[T]]
```

Comment:
I believe that in fact we will never put a `CascadingPipe` directly into this cache. We could assert that to make it clear. It could come in as a grouping from someone that uses `from(pipe: Pipe, f: Field)...`. That said, in that case, I think we want identity hashCode and equals. I think we can add a discussion of this in the comment.